pkg/kubelet/config/file.go

```diff
@@ -62,26 +65,32 @@ func new(path string, nodeName types.NodeName, period time.Duration, updates cha
 		updates <- kubetypes.PodUpdate{Pods: pods, Op: kubetypes.SET, Source: kubetypes.FileSource}
 	}
 	store := cache.NewUndeltaStore(send, cache.MetaNamespaceKeyFunc)
-	return &sourceFile{
+	return &sourceFileListerWatcher{
 		path:           path,
 		nodeName:       nodeName,
+		period:         period,
 		store:          store,
 		fileKeyMapping: map[string]string{},
 		updates:        updates,
+		locker:         &sync.Mutex{},
 	}
 }
 
-func (s *sourceFile) run() {
-	if err := s.watch(); err != nil {
-		glog.Errorf("unable to read config path %q: %v", s.path, err)
-	}
+func (s *sourceFileListerWatcher) run() {
```

Comment:
I ll take `run & watch` you said as `list & watch`.

Period `list` is added to fully scan files. Some special cases can not be not covered by inotify.

For example, suppose there is a symlink under the watched dir `meta_dir` which is linking a non-existing file. And the linked file is then created, after our `sourceFileWatcher` starts. The `sourceFileWatcher` only using inotify is not able to detect it. This case is tested by `file_linux_test.go.watchFileAdded` with `watchDir=true` and `symlink=true`. If commenting the `list` process, event after adding `IN_ATTRIB`, this case will fail.